File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };

To reverse a link list is quite straightforward 
Here is an example
    n1 -> n2 -> null
    Will become
    n2 -> n1 -> null

    The idea is to reverse the list and have two pointers the previous and the current
    1. Save the next pointer of the current. (nextSave = cur -> next), 
    The first step is very essential because we ensure after 
    reverse we still keep track of the original next node so this has to be done in the first 
    and cannot be done after the second step otherwise we will loop to ourself back

    2. Then then move the current next to the previous (cur -> next = prev)


    3. Find the above step two and three ensure the pointer is reversed.

    4. Then we move our group completely forward by the following
        5-1: prev = cur
        5-2: cur = nextSave

For complexity analysis
    The time complexity will be linear because we illiterate the LL at once
    Space complexity will be constant because there is no extra linear space to store the data

 */


class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* cur = head;
        ListNode* prev = nullptr;
        ListNode* savedNext = nullptr;

        while(cur) {
            // saved next
            savedNext = cur->next;
            // reverse
            cur->next = prev;

            // move group forward
            prev = cur;
            cur = savedNext;
        }

        return prev;
    }
};

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

/*
Intuition: 
	2 pass to use the hashmap to store the next node for random and next

	,a
*/
func copyRandomList(head *Node) *Node {
	if head == nil {
		return nil
	}
	// store the next pointer
	oldToNew := make(map[*Node]*Node) // maps old to new node
	newHead := &Node{Val: head.Val}
	oldToNew[head] = newHead

	prev := newHead
	cur := head.Next

	for cur != nil {
		prev.Next = &Node{
			Val: cur.Val,
		}

		// map and move both
		oldToNew[cur] = prev.Next
		cur = cur.Next
		prev = prev.Next
	}

	// Second pass: set random pointers
	cur = head
	curNew := newHead
	for cur != nil {
    	curNew.Random = oldToNew[cur.Random]
    	curNew = curNew.Next
    	cur = cur.Next
	}

	return newHead
}
